src/project/git/StableDirectoryManager.ts

```diff
@@ -0,0 +1,135 @@
+}
+
+const DefaultOpts = {
+
+    maxDirectories: 100,
+    cleanOnExit: true,
+};
+
+/**
+ * Build a stable working directory structure.
+ * Options determine whether it's cleared on exit.
+ */
+export class StableDirectoryManager implements DirectoryManager {
+
+    public opts: StableDirectoryManagerOpts;
+
+    private directoriesUsed = 0;
+
+    constructor(pOpts: StableDirectoryManagerOpts) {
+        this.opts = {
+            ...DefaultOpts,
+            ...pOpts,
+        };
+
+        logger.info("Setting up StableDirectoryManager with options %s", JSON.stringify(this.opts));
+
+        if (this.opts.cleanOnExit === true) {
+            logger.info("Registering shutdown hook to delete [%s]", this.opts.baseDir);
+            registerShutdownHook(() => {
+                logger.info("Cleaning up temporary directories under [%s]", this.opts.baseDir);
+                return fs.remove(this.opts.baseDir).then(() => 0);
+            });
+        }
+    }
+
+    public directoryFor(user: string, repo: string, branch: string, opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        if (this.opts.reuseDirectories) {
+            // Attempt to reuse the directory
+            return this.existingDirectoryFor(user, repo, branch, opts)
+                .then(existing =>
+                    !!existing ? existing : this.freshDirectoryFor(user, repo, branch, opts));
+        } else {
+            return this.freshDirectoryFor(user, repo, branch, opts);
+        }
+    }
+
+    /**
+     * Return undefined if not found
+     */
+    private existingDirectoryFor(user: string, repo: string, branch: string,
+                                 opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        const expectedPath = this.pathFor(user, repo);
+        return fs.pathExists(expectedPath)
+            .then(exists => {
+                if (exists) {
+                    logger.info("%s directories used of %s: Reusing path [%s]",
+                        this.directoriesUsed, this.opts.maxDirectories, expectedPath);
+                    return {
+                        path: expectedPath,
+                        type: "actual-directory" as any,
+                    };
+                } else {
+                    // It doesn't exist
+                    return undefined;
+                }
+            });
+    }
+
+    private freshDirectoryFor(user: string, repo: string, branch: string,
+                              opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        return this.createDir(user, repo)
+            .then(path => {
+                this.directoriesUsed++;
+                logger.info("%s directories used of %s: Returning new path [%s]",
+                    this.directoriesUsed, this.opts.maxDirectories, path);
+                return {
+                    path,
+                    type: "parent-directory" as any,
+                };
+            });
+    }
+
+    private pathFor(user: string, repo: string): string {
```

Comment:
all this "user" is really "owner" right?

Reply:
Yeah